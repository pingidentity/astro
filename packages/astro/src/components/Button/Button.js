import React, { forwardRef, useRef, useImperativeHandle } from 'react';
import PropTypes from 'prop-types';
import { mergeProps, useButton, useFocusRing } from 'react-aria';
import { Button as ThemeUIButton } from 'theme-ui';
import { useHover } from '@react-aria/interactions';

import { useStatusClasses, usePropWarning, useAriaLabelWarning, useDeprecationWarning } from '../../hooks';
import Loader from '../Loader';

/**
 * Buttons are used to trigger actions or events.
 * This component is based on the [Button - Theme-UI](https://theme-ui.com/components/button/)
 * and includes a variety of styles, such as primary action, secondary action, or warning.
 */

const Button = forwardRef((props, ref) => {
  const {
    className,
    isDisabled,
    isLoading,
    onHoverStart,
    onHoverChange,
    onHoverEnd,
    onPress,
    onPressStart,
    onPressEnd,
    onPressChange,
    onPressUp,
    children,
    variant,
    tabIndex,
    ...others
  } = props;
  const buttonRef = useRef();
  usePropWarning(props, 'disabled', 'isDisabled');
  /* istanbul ignore next */
  useImperativeHandle(ref, () => buttonRef.current);

  const { isFocusVisible, focusProps } = useFocusRing();
  const { buttonProps, isPressed } = useButton({
    elementType: 'button',
    ...props,
  }, buttonRef);
  const { hoverProps, isHovered } = useHover(props);
  const { classNames } = useStatusClasses(className, {
    isHovered,
    isPressed,
    isFocused: isFocusVisible,
    isDisabled,
  });

  useDeprecationWarning(
    'The "icon" variant for `Button` is deprecated in Astro-UI 1.0.0, use the `IconButton` component instead.',
    { isActive: props.variant === 'icon' },
  );

  useDeprecationWarning(
    'The "danger" variant for `Button` will be deprecated in Astro-UI 2.0.0, use the `critical` variant instead.',
    { isActive: props.variant === 'danger' },
  );

  useDeprecationWarning(
    'The "text" variant for `Button` will be deprecated in Astro-UI 2.0.0, use the `link` variant instead.',
    { isActive: props.variant === 'text' },
  );

  const ariaLabel = props['aria-label'];
  useAriaLabelWarning('Button', ariaLabel, variant === 'filter');

  return (
    <ThemeUIButton
      aria-label={ariaLabel}
      ref={buttonRef}
      className={classNames}
      role="button"
      tx="buttons"
      sx={isLoading && { display: 'flex', justifyContent: 'center', alignItems: 'center' }}
      variant={variant}
      {...others}
      {...mergeProps({ ...buttonProps, tabIndex }, hoverProps, focusProps)}
    >
      {isLoading ? <span style={{ visibility: 'hidden' }}>{children}</span> : children}
      {isLoading && <Loader size="0.5em" sx={{ position: 'absolute' }} />}
    </ThemeUIButton>
  );
});

Button.propTypes = {
  /** Defines a string value that labels the current element. */
  'aria-label': PropTypes.string,
  /** Whether the button is disabled. */
  isDisabled: PropTypes.bool,
  /** Shows loader instead of children */
  isLoading: PropTypes.bool,
  /**
   * Handler that is called when a hover interaction starts.
   * (e: HoverEvent) => void
   */
  onHoverStart: PropTypes.func,
  /**
   * Handler that is called when a hover interaction ends.
   * (e: HoverEvent) => void
   */
  onHoverEnd: PropTypes.func,
  /**
   * Handler that is called when the hover state changes.
   * (isHovering: boolean) => void
   */
  onHoverChange: PropTypes.func,
  /**
   * Handler that is called when the press is released over the target.
   * (e: PressEvent) => void
   */
  onPress: PropTypes.func,
  /**
   * Handler that is called when a press interaction starts.
   * (e: PressEvent) => void
   */
  onPressStart: PropTypes.func,
  /**
   * Handler that is called when a press interaction ends, either over the target or when the
   * pointer leaves the target.
   * (e: PressEvent) => void
   */
  onPressEnd: PropTypes.func,
  /**
   * Handler that is called when the press state changes.
   * (isPressed: boolean) => void
   */
  onPressChange: PropTypes.func,
  /**
   * Handler that is called when a press is released over the target, regardless of whether it
   * started on the target or not.
   * (e: PressEvent) => void
   */
  onPressUp: PropTypes.func,
  /** The styling variation of the button. */
  variant: PropTypes.string,
  /** The focus variation of the button. */
  tabIndex: PropTypes.number,
};

Button.defaultProps = {
  isDisabled: false,
  variant: 'default',
};

Button.displayName = 'Button';
export default Button;
