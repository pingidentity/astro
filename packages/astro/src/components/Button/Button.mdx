<Meta title="Components/Button/Button" />

# Button

Button is used primarily to trigger an action or event. It is built on top of
[Button from Theme-UI](https://theme-ui.com/components/button). The stylings of a button is different based on its purpose.
(Primary Action, Secondary Action, Danger Button)

### Required components 

- Button component can be used independently and does not require additional components.

### Best practices

Button should:
- be clearly and accurately labeled.

Button shouldnâ€™t:
- be used as links.
- override colors (accessibility purposes).
- use uppercase titles.

### Accessibility

This component should adhere to the [WAI-ARIA Button](https://www.w3.org/TR/wai-aria-practices-1.1/examples/button/button.html) accessibility guidelines.

#### Keyboard Navigation

- Buttons are focusable using the Tab key and can be selected using the Space or Enter key. 

#### Screen readers.

Button utilizes `aria-label` to give an accessible name to the component.

| Aria Attribute | Role | Component |
| ---- | ---- | ----------- |
| `aria-label` | button | Button |