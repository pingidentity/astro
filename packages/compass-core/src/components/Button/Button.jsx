import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import { noop } from 'underscore';
import { useFocusOutline } from '../../utils/FocusUtils';

/**
 * Basic button component.
 */

const Button = forwardRef(({
    children,
    className,
    css,
    'data-id': dataId,
    isDisabled,
    href,
    onClick,
    onMouseDown,
    style,
    isSubmit,
    target,
    ...props
}, ref) => {
    useFocusOutline();

    const TagName = href ? 'a' : 'button';

    return (
        <TagName
            className={className}
            css={css}
            data-id={dataId}
            onClick={onClick}
            onMouseDown={onMouseDown}
            ref={ref}
            disabled={isDisabled}
            type={isSubmit ? 'submit' : 'button'}
            href={href}
            style={style}
            target={target}
            {...props}
        >
            {children}
        </TagName>
    );
});

Button.displayName = 'Button';

Button.propTypes = {
    /** Extra CSS class(s) applied to the top-level HTML container. */
    className: PropTypes.string,
    /** CSS prop can be used for CSS-in-JS components. */
    css: PropTypes.object,
    /** Defines the "data-id" for top-level HTML container. */
    'data-id': PropTypes.string,
    /** Button will not function when true. */
    isDisabled: PropTypes.bool,
    /** Pass in a string to give the button an href. Converts to an a tag for a11y. */
    href: PropTypes.string,
    /** Click handler.
     * @param {object} event
    */
    onClick: PropTypes.func,
    /** Mouse down handler.
     * @param {object} event
    */
    onMouseDown: PropTypes.func,
    /** Set inline styles. */
    style: PropTypes.shape({}),
    /** Set to true to render a submit type button. */
    isSubmit: PropTypes.bool,
    /** Pass a string to give the button a target. Only relevant if it also has href. */
    target: PropTypes.string,
};

Button.defaultProps = {
    isDisabled: false,
    onClick: noop,
    onMouseDown: noop,
    isSubmit: false,
};

export default Button;
