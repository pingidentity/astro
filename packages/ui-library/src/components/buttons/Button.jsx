import React from "react";
import PropTypes from "prop-types";
import classnames from "classnames";
import HelpHint from "../tooltips/HelpHint";
import EllipsisLoader from "../general/EllipsisLoader";
import { cannonballChangeWarning } from "../../util/DeprecationUtils";
import { flagsPropType } from "../../util/FlagUtils";

/**
 * @class Button
 * @desc button component
 *
 * @param {("primary" | "secondary" | "standard" | "success" | "cancel" | "danger")} [type]
 *      CSS class applied to html.
 * @param {string} [alignInputs]
 *     When true, aligns a  button with text fields in the same row
 * @param {string} [className]
 *     Extra CSS class(s) applied to the top-level HTML container.
 * @param {string} [data-id="button"]
 *     Defines the "data-id" for top-level HTML container.
 * @param {string} [iconName]
 *     The name and CSS of the icon.
 * @param {function} [onClick]
 *     Click handler.
 * @param {string} [label]
 *     Html name of the button.
 * @param {boolean} [disabled=false]
 *     Button will not function when true.
 * @param {string} [disabledText]
 *     Text for the help hint will be rendered when the same button has a the prop disabled set to true.
 * @param {boolean} [inline]
 * @param {boolean} loading
 *     While true, loading animation will be shown.
 *  @param {boolean} [noSpacing]
 *     Don't include the right margin
 * @param {boolean} active
 *     Active style of the button for when it's being used as a toggle.
 * @param {array} [flags]
 *     Set the flag for "add-button-margin" to override the special add button margin styling.
 * @example
 *  <Button
 *      label="Button"
 *      type="Danger">
 *  </Button>
 *
 */

const buttonTypes = {
    PRIMARY: "primary",
    SECONDARY: "secondary",
    STANDARD: "standard",
    SUCCESS: "success",
    CANCEL: "cancel",
    DANGER: "danger"
};

const buttonTypeList = Object.values(buttonTypes);

const dontFocus = event => event.preventDefault();

const checkFixAddMargin = flags => flags.includes("add-button-margin");

function Button ({
    active,
    alignInputs,
    children,
    className,
    "data-id": dataId,
    disabled,
    disabledText,
    flags,
    href,
    iconName,
    inline,
    label,
    loading,
    noSpacing,
    onClick,
    submit,
    target,
    type
}) {
    const fixAddMargin = checkFixAddMargin(flags);

    if (iconName === "add" && !fixAddMargin && !noSpacing) {
        cannonballChangeWarning({
            message: (
                `Add buttons will no longer automatically have their right margin stripped. ` +
                    `Instead, the 'noSpacing' prop will manually do it. ` +
                    `Switch over by adding the 'add-button-margin' flag to Add buttons.`
            ),
        });
    }

    const classes = classnames(
        "button",
        className,
        iconName,
        {
            "align-inputs": alignInputs,
            "button--add-margin-fix": iconName === "add" && fixAddMargin,
            "button--no-spacing": noSpacing,
            inline,
            loading: loading,
            disabled: disabled,
            "button--active": active,
            "ellipsis-loader-button": loading,
            [type]: buttonTypeList.includes(type)
        }
    );

    const TagName = href ? "a" : "button";

    const Tags = (
        <TagName
            className = {classes}
            data-id={dataId}
            onMouseDown={dontFocus}
            onClick={onClick}
            disabled={disabled}
            type={submit ? "submit" : "button"}
            href={href}
            target={target}
        >
            {label}
            {children}
            <EllipsisLoader loading={loading}/>
        </TagName>
    );

    return (
        disabled && disabledText
            ? <HelpHint
                data-id="button_help-hint"
                hintText={disabledText}
                placement="top"
            >
                {Tags}
            </HelpHint>
            : Tags
    );
}

Button.propTypes = {
    active: PropTypes.bool,
    alignInputs: PropTypes.bool,
    className: PropTypes.string,
    "data-id": PropTypes.string,
    disabled: PropTypes.bool,
    disabledText: PropTypes.string,
    flags: flagsPropType,
    href: PropTypes.string,
    iconName: PropTypes.string,
    inline: PropTypes.bool,
    label: PropTypes.node,
    loading: PropTypes.bool,
    noSpacing: PropTypes.bool,
    onClick: PropTypes.func,
    submit: PropTypes.bool,
    target: PropTypes.string,
    type: PropTypes.oneOf(buttonTypeList),
};

Button.defaultProps = {
    active: false,
    alignInputs: false,
    "data-id": "button",
    disabled: false,
    flags: [],
    loading: false,
    noSpacing: false,
    onClick: () => {},
    submit: false,
};

Button.buttonTypes = buttonTypes;

export default Button;
