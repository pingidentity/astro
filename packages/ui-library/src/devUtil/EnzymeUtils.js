import { mount } from "enzyme";


export const snapshotDataIds = node => {
    const elements = node.find("[data-id]").hostNodes(); // Find only actual HTML elements

    const ids = elements.map(element => element.prop("data-id")).filter(item => item !== "");
    expect(ids).toMatchSnapshot();
};

export const mountSnapshotDataIds = node => snapshotDataIds(mount(node));
