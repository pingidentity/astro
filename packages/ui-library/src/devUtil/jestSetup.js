
// This is needed for testing the modal and DetailsTooltip component.
global.getSelection = jest.fn();
global.getSelection.mockReturnValue({
    toString: () => "",
});
global.document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
        nodeName: "BODY",
        ownerDocument: document,
    },
});
