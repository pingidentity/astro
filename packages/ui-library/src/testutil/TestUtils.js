/*eslint-disable valid-jsdoc*/

"use strict";

/**
 * @module util/TestUtils
 * @desc A collection of ReactJS test utils, to complement the ones provided by the
 * framework. This is a copy of the TestUtils in the library, just because this class
 * is used in the tests and should work with the same version of React used in the WP
 * and not in the library.
 */

var ReactDOM = require("react-dom");
var React = require("react");
var ReactTestUtils = require("react-addons-test-utils");
var _ = require("underscore");
var assign = require("object-assign");

var TestUtils = {

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithDataId
     * @desc Return all DOM nodes children of the given tree with the specified dataId.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} dataId - the value of the data-id attribute on the target elements
     * @return {array} - of DOM node objects which match the given criteria
     */
    scryRenderedDOMNodesWithDataId: function (parent, dataId) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.getAttribute("data-id") === dataId) {
            nodes = [parentNode];
        }
        else {
            var selectorValue = dataId
                .replace(/\[/g, "\\[")
                .replace(/\]/g, "\\]")
                .replace(/'/g, "\\'");
            nodes = parentNode.querySelectorAll("[data-id=" + selectorValue + "]");
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithDataId
     * @desc Return the single DOM node child of the given tree with the specified dataId.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} dataId - the value of the data-id attribute on the target element
     * @return {object} - the DOM node object which matches the given criteria
     */
    findRenderedDOMNodeWithDataId: function (parent, dataId) {
        var nodes = this.scryRenderedDOMNodesWithDataId(parent, dataId);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with dataId: " +
                dataId + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithTag
     * @desc Return all DOM nodes children of the given tree with the specified tag name.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} tagName - the tag name on the target elements
     * @return {array} - array of DOM nodes object which match the given criteria
     */
    scryRenderedDOMNodesWithTag: function (parent, tagName) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.tagName.toLowerCase() === tagName.toLowerCase()) {
            nodes = [parentNode];
        }
        else {
            nodes = parentNode.querySelectorAll(tagName);
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithTag
     * @desc Return the single DOM node child of the given tree with the specified tag name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} tagName - the tag name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithTag: function (parent, tagName) {
        var nodes = this.scryRenderedDOMNodesWithTag(parent, tagName);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with tag name: " +
                tagName + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithName
     * @desc Return all DOM nodes children of the given tree with the specified name.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} name - the name on the target elements
     * @return {DOMNode[]} - all DOM nodes which match the given criteria
     */
    scryRenderedDOMNodesWithName: function (parent, name) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.name === name) {
            nodes = [parentNode];
        }
        else {
            nodes = parentNode.querySelectorAll("[name='" + name.replace("'", "\\'") + "']");
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithName
     * @desc Return the single DOM node child of the given tree with the specified name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} name - the name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithName: function (parent, name) {
        var nodes = this.scryRenderedDOMNodesWithName(parent, name);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with name: " +
                name + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithClass
     * @desc Return all DOM nodes children of the given tree with the specified CSS class name.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} className - the CSS class name on the target elements (or a concatenation of classes, separated by space)
     * @return {DOMNode[]} - all DOM nodes which match the given criteria
     */
    scryRenderedDOMNodesWithClass: function (parent, className) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode = ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
            ? ReactDOM.findDOMNode(parent)
            : parent;

        var nodes;

        // first check if the root element has the provided classes;
        // if not, build a selector and look for any children

        var parentMatches = _.every(className.split(" "), function (name) {
            return (" " + parentNode.className + " ").indexOf(" " + name + " ") > -1;
        });
        if (parentMatches) {
            nodes = [parentNode];
        }
        else {
            var selector = _.reduce(className.split(" "), function (memo, name) {
                return memo + "." + name;
            }, "");
            nodes = parentNode.querySelectorAll(selector);
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithClass
     * @desc Return the single DOM node child of the given tree with the specified CSS class name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} className - the CSS class name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithClass: function (parent, className) {
        var nodes = this.scryRenderedDOMNodesWithClass(parent, className);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with class name: " +
                className + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedComponentsWithType
     * @desc Finds all instances of components with type equal to `componentType`.
     * @param {object} root - a composite (ie. React) component
     * @param {object} componentType - the component type
     * @return {array} an array of all the matches.
     */
    scryRenderedComponentsWithType: function (root, componentType) {
        return ReactTestUtils.scryRenderedComponentsWithType(root, componentType);
    },

    /**
     * @alias module:util/TestUtils.findRenderedComponentWithType
     * @desc Same as `scryRenderedComponentsWithType` but expects there to be one result
     * and returns that one result, or throws exception if there is any other
     * number of matches besides one.
     * @param {object} root - a composite (ie. React) component
     * @param {object} componentType - the component type
     * @return {!ReactComponent} The one match.
     */
    findRenderedComponentWithType: function (root, componentType) {
        return ReactTestUtils.findRenderedComponentWithType(root, componentType);
    },

    /**
     * @alias module:util/TestUtils.captureGlobalListener
     * @desc Attaches proxy function for `window.addEvenListener` which will save reference to actual provided callback when
     * event type match filter. Useful for testing ReactJS component global (window) callbacks behaviour.
     *
     * @param {string} eventType event to save callback for. case insensitive.
     * @param {object} [obj=window] optional object to capture listener from. Like `document`. If not provided - window object
     * @returns {function} - when executed will trigger actual provided callback to `window.addEventListener`
     *
     * @example
     * //capture registered callback from reactjs control
     * var onKeyDown = TestUtils.captureGlobalListener("keyDown");
     *
     * onKeyDown({keyCode:13});  //later in test trigger global event
     *
     * @example
     * //capture registered callback from reactjs control on `document` object
     * var onKeyDown = TestUtils.captureGlobalListener("keyDown", document);
     *
     * onKeyDown({keyCode:13});  //later in test trigger global event
     */
    captureGlobalListener: function (eventType, obj) {

        var captured;

        var target = obj || window;

        target.addEventListener = function (event, listener) {
            if (event.toLowerCase() === eventType.toLowerCase()) {
                captured = listener;
            }
        };

        return function () {
            return captured.apply(this, arguments);
        };
    }

    /**
     * @desc Return all components in the supplied tree with the specified dataId.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} dataId - the value of the data-id attribute on the target elements
     * @return {ReactComponent[]} - all components which match the given criteria
     */
    /*
     scryRenderedDOMComponentsWithDataId: function (tree, dataId) {
     var components = ReactTestUtils.findAllInRenderedTree(tree, function (inst) {
     var instId = ReactDOM.findDOMNode(inst).getAttribute("data-id");
     return ReactTestUtils.isDOMComponent(inst) && (instId === dataId);
     });

     return components;
     },
     */

    /**
     * @desc Return the single component in the supplied tree with the specified dataId.
     * If no component in the tree matches the criteria, null is returned.
     * If there are more than one component, an error is thrown.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} dataId - the value of the data-id attribute on the target element
     * @return {ReactComponent} - the component which match the given criteria
     * @throws Will throw an Error if more than one element match the criteria
     */
    /*
     findRenderedDOMComponentWithDataId: function (tree, dataId) {
     var components = TestUtils.scryRenderedDOMNodesWithDataId(tree, dataId);

     if (components.length > 1) {
     throw new Error("Found more than one component with dataId: " +
     dataId + " in the given tree");
     }

     return components.length === 0 ? null : components[0];
     },
     */

    /**
     * @desc Return the single component in the supplied tree with the specified name.
     * If no component in the tree matches the criteria, null is returned.
     * If there are more than one component, an error is thrown.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} name - the value of the name attribute on the target element
     * @return {ReactComponent} - the component which match the given criteria
     * @throws Will throw an Error if more than one element match the criteria
     */
    /*
     findRenderedDOMComponentWithName: function (tree, name) {
     var components = ReactTestUtils.findAllInRenderedTree(tree, function (inst) {
     var instName = ReactDOM.findDOMNode(inst).name;
     return ReactTestUtils.isDOMComponent(inst) && (name === instName);
     });

     if (components.length > 1) {
     throw new Error("Found more than one component with name: " +
     name + " in the given tree");
     }

     return components.length === 0 ? null : components[0];
     }
     */
};

/**
 * @class UpdatePropsWrapper
 *
 * @desc
 * A simple wrapper to be used for unit testing, when you need to update
 * a property on a rendered component. Starting with React 0.14.x,
 * directly setting or replacing a prop on a rendered component is deprecated.
 * Instead, use this wrapper and call _setProps(newProps) on it
 * to have the wrapped component re-render with the new property(ies).
 * The ref attribute on the wrapper is set to "wrapper".
 *
 * @Example
 *  // initial render
 *  var component = ReactTestUtils.renderIntoDocument(
 *      <UpdatePropsWrapper type={If} test={true} />
 *  );
 *  // update the property through re-render
 *  component._setProps({test: false});
 *
 * @Example
 *  var component = ReactTestUtils.renderIntoDocument(
 *      <UpdatePropsWrapper type={TabbedSections}
 *              onSectionChange={jest.genMockFunction()}
 *              selectedIndex={-1}>
 *          <div data-id="section1" title="section 1">section 1</div>
 *          <div data-id="section2" title="section 2">section 2</div>
 *      </UpdatePropsWrapper>
 *  );
 *  component._setProps({ selectedIndex: 0 });
 *
 */
TestUtils.UpdatePropsWrapper = React.createClass({
    /**
     * Re-render the wrapped component using the new property.
     * Use this function to avoid setting the props directly on the component (anti pattern).
     * @param {object} props - new property(ies) to set
     */
    _setProps: function (props) {
        this.setState(props);
    },
    getInitialState: function () {
        return assign({ ref: "wrapper" }, this.props);
    },
    render: function () {
        return React.createElement(
            this.props.type, // this is the type of element to create, provided as prop
            this.state // this is the rest of props to set on the element
        );
    }
});

module.exports = TestUtils;
