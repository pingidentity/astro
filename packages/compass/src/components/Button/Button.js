import React from 'react';
import PropTypes from 'prop-types';
import CoreButton from '@pingux/compass-core/lib/components/Button';
import PlusSVG from '@mdi/svg/svg/plus.svg';
import CheckSVG from '@mdi/svg/svg/check-bold.svg';
import ArrowRightSVG from '@mdi/svg/svg/arrow-right.svg';
import AlertSVG from '@mdi/svg/svg/alert.svg';
import { typeToVariant } from '@pingux/compass-core/lib/utils/PropUtils';

import { makeIcon } from '../Icon';
import { textProps } from '../../styles/text';
import ButtonLoader from './ButtonLoader';

import useCompassTheme from '../../styles/useCompassTheme';
import { buttonStyle, statusIconStyle, buttonIconStyle } from './Button.styles';
import { buttonStatuses, buttonVariants } from './Button.constants';

const {
    PRIMARY,
} = buttonVariants;

const {
    CRITICAL,
    LOADING,
    SUCCESS,
} = buttonStatuses;

const buttonIcons = {
    'alert': makeIcon(AlertSVG, 'Alert'),
    'arrow-right': makeIcon(ArrowRightSVG, 'Right Arrow'),
    'plus': makeIcon(PlusSVG, 'Plus'),
    'check': makeIcon(CheckSVG, 'Check'),
};

const renderButtonIcon = (key, buttonProps) => {
    const ButtonIcon = buttonIcons[key];
    const { isInline } = buttonProps;

    return (
        <ButtonIcon
            css={buttonIconStyle(buttonProps)}
            bg={!isInline ? 'white' : undefined}
            mx={isInline ? -5 : undefined}
        />
    );
};

const renderStatusIcon = (key, buttonProps) => {
    const {
        theme: { colors },
        variant,
        status,
    } = buttonProps;
    const isPrimary = variant === PRIMARY;
    const isCritical = status === CRITICAL;

    if (key === 'loader') {
        return (
            <ButtonLoader color={isPrimary ? colors.white : colors.active} />
        );
    }

    const ButtonIcon = buttonIcons[key];
    return (
        <ButtonIcon
            css={statusIconStyle(buttonProps)}
            bg={isCritical ? undefined : 'white'}
        />
    );
};

const Button = ({
    children,
    icon,
    iconAfter,
    isInline,
    status,
    type, // eslint-disable-line react/prop-types
    variant = typeToVariant('Button', Object.values(buttonVariants), type, 'default'),
    ...props
}) => {
    const theme = useCompassTheme();
    const themeProps = {
        theme,
        isInline,
        status,
        variant,
        ...props,
    };

    return (
        <CoreButton
            css={buttonStyle({
                ...textProps({ weight: 1, size: isInline ? 'sm' : 'md' }),
                ...themeProps,
            })}
            {...props}
        >
            {icon && renderButtonIcon(icon, themeProps)}
            {status === SUCCESS && renderStatusIcon('check', themeProps)}
            {status === CRITICAL && renderStatusIcon('alert', themeProps)}
            {status === LOADING && renderStatusIcon('loader', themeProps)}
            <span>{children}</span>
            {iconAfter && renderButtonIcon(iconAfter, themeProps)}
        </CoreButton>
    );
};

Button.propTypes = {
    /** Icon that appears on the left of the label */
    icon: PropTypes.oneOf(Object.keys(buttonIcons)),
    /** Icon that appears on the right of the label */
    iconAfter: PropTypes.oneOf(Object.keys(buttonIcons)),
    /** Button will not function when true. */
    isDisabled: PropTypes.bool,
    /** Triggers the pill button style */
    isInline: PropTypes.bool,
    /** Status of the button for feedback purposes */
    status: PropTypes.oneOf(Object.values(buttonStatuses)),
    /** Basic appearance of the button */
    variant: PropTypes.oneOf(Object.values(buttonVariants)),
};

Button.defaultProps = {
    isDisabled: false,
    isInline: false,
    status: 'normal',
};

export default Button;
