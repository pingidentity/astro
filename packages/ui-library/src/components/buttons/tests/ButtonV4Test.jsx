window.__DEV__ = true;

jest.dontMock("../Button");

import React from "react";
import ReactTestUtils from "react-dom/test-utils";
import _ from "underscore";
import TestUtils from "../../../testutil/TestUtils";
import { mountSnapshotDataIds } from "../../../devUtil/EnzymeUtils";
import Button from "../Button";

describe("Button", function () {
    const componentId = "button";

    function getComponent (opts) {
        opts = _.defaults(opts || {}, {
            "data-id": componentId,
        });

        return ReactTestUtils.renderIntoDocument(<div><Button {...opts} /></div>);
    }

    it("rendered component with data-id=button", function () {
        const component = getComponent({
            submit: true
        });

        const element = TestUtils.findRenderedDOMNodeWithDataId(component, componentId);

        expect(ReactTestUtils.isDOMComponent(element)).toBeTruthy();
    });

    it("data-id's do not change", () => {
        mountSnapshotDataIds(
            <Button
                disabled
                disabledText="test"
            />
        );
    });

    it("renders href tag with 'a' ", function () {
        const component = getComponent({
            href: "cnn.com"
        });

        const element = TestUtils.findRenderedDOMNodeWithTag(component, "a");

        expect(ReactTestUtils.isDOMComponent(element)).toBeTruthy();
    });

    it("renders a help hint if disabconstext has a string and disabled is true ", function () {
        const component = getComponent({
            disabledText: "hello",
            disabled: true
        });

        const element = TestUtils.findRenderedDOMNodeWithDataId(component, "button_help-hint");

        expect(ReactTestUtils.isDOMComponent(element)).toBeTruthy();
    });


    it("renders the button with click callback", function () {
        const onClick = jest.fn();
        const component = getComponent({
            onClick: onClick
        });
        const button = TestUtils.findRenderedDOMNodeWithDataId(component, componentId);

        expect(onClick).not.toBeCalled();
        ReactTestUtils.Simulate.click(button);
        expect(onClick).toBeCalled();
    });

    it("does not throw error with default onClick", () => {
        const component = getComponent();
        const button = TestUtils.findRenderedDOMNodeWithDataId(component, componentId);

        ReactTestUtils.Simulate.click(button);
        expect(button).toBeTruthy();
    });
});
