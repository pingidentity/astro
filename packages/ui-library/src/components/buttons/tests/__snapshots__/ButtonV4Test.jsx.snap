// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button data-id's do not change 1`] = `
Array [
  "button_help-hint",
  "button_help-hint-target",
  "button",
  "tooltip",
]
`;
